File: Commands/CommandIterator.java
package Commands;

import java.util.List;

/**
 * Created by Mark Chimes on 2017/11/05.
 */
public class CommandIterator implements CommandOrMessageIterator {
    private final List<String> items;
    private int iteratorIndex;

    public CommandIterator(List<String> items) {
        this(items, 0);
    }

    public CommandIterator(List<String> items, int index) {
        this.items = items;
        iteratorIndex = index;
    }

    public String currentText() {
        return items.get(iteratorIndex);
    }

    public int currentIndex() { return iteratorIndex; }

    @Override
    public void performPreviousInListCommand() {
        if (iteratorIndex > 0) {
            iteratorIndex--;
        } else {
            iteratorIndex = items.size() - 1;
        }
    }

    @Override
    public void performNextInListCommand() {
        if (iteratorIndex < items.size() - 1) {
            iteratorIndex++;
        } else {
            iteratorIndex = 0;
        }
    }
}

File: commandAndMessage/command/CommandIterator.kt
package commandAndMessage.command

/**
 * Created by Mark Chimes on 2017/11/05.
 */
class CommandIterator @JvmOverloads constructor(private val items: List<CommandTuple>, private var iteratorIndex: Int = 0) : CommandOrMessageIterator {

    override fun currentText(): String {
        return currentCommand().actionString
    }

    fun currentCommand(): CommandTuple {
        return items[iteratorIndex]
    }

    override fun currentIndex(): Int {
        return iteratorIndex
    }

    override fun performPreviousInListCommand() {
        if (iteratorIndex > 0) {
            iteratorIndex--
        } else {
            iteratorIndex = items.size - 1
        }
    }

    override fun performNextInListCommand() {
        if (iteratorIndex < items.size - 1) {
            iteratorIndex++
        } else {
            iteratorIndex = 0
        }
    }
}